### ISSUE TITLE
Ubuntu 17.10: nvidia-docker2 : Depends: docker-ce (= 17.12.0~ce-0~ubuntu) but it is not installable

### ISSUE THREAD

**Issue by jwatte** [NVIDIA/nvidia-docker#643]
I need to run Nvidia Jetpack to upgrade my Jetson board.
My workstation runs Ubuntu 17.10.
Jetpack installer requires Ubuntu 16.04. (I can flash from the command line from 17.10, but the actual installer requires 16.04)
So, I want to run the Jetpack installer on 16.04 using nvidia-docker.

I follow the instructions in the README on the nvidia-docker github, but get the following error:

```
[10:48] jwatte@ripper:~/xenial$ sudo apt-get install -y nvidia-docker2
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 nvidia-docker2 : Depends: docker-ce (= 17.12.0~ce-0~ubuntu) but it is not installable or
                           docker-ee (= 17.12.0~ee-0~ubuntu) but it is not installable
E: Unable to correct problems, you have held broken packages.

```

The version of docker.io installed on my workstation is:

```
[10:51] jwatte@ripper:~/xenial$ dpkg -l docker.io
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                             Version               Architecture          Description
+++-================================-=====================-=====================-=====================================================================
ii  docker.io                        1.13.1-0ubuntu6       amd64                 Linux container runtime
[10:51] jwatte@ripper:~/xenial$ dpkg -l docker-ce
dpkg-query: no packages found matching docker-ce

```

The compatibility matrix says that docker.io is a supported version. (This is the package available for Ubuntu; docker-ce/ee is not.)

**Comment by flx42** [comment]
You need to pin the version of `nvidia-docker2` and `nvidia-container-runtime`:
https://github.com/NVIDIA/nvidia-docker/wiki/Frequently-Asked-Questions#how-do-i-install-20-if-im-not-using-the-latest-docker-version

It's a bit complicated today, we'll see if we can improve this step for `docker.io`.

**Comment by YoungCholKim** [comment]
I've installed the latest version of docker`Version:	18.02.0-ce` but failed with the same error:
`nvidia-docker2 : Depends: docker-ce (= 17.12.0~ce-0~ubuntu) but 18.02.0~ce-0~ubuntu is to be installed or
                           docker-ee (= 17.12.0~ee-0~ubuntu) but it is not installable
`

I believe the Prerequisites doc `Docker >= 1.12`   or  version check codes  should be changed.

**Comment by jwatte** [comment]
Yes, the solution for me was to uninstall `docker.io` and `docker-ce`, and then install `docker-ce=17.12.0~ce-0~ubuntu`
The documentation that says `docker.io` is supported should probably also be updated to remove that reference.

**Comment by flx42** [comment]
No, it is supported, but it's just ugly until we add a new virtual package:
```
sudo apt-get install -y docker.io nvidia-container-runtime=1.1.1+docker1.13.1-1 nvidia-docker2=2.0.2+docker1.13.1-1
```

**Comment by shrx** [comment]
@flx42 I get this error:
```
 nvidia-docker2 : Depends: docker-engine (= 1.13.1-0~ubuntu-xenial) but it is not installable or
                           docker.io (= 1.13.1-0ubuntu1~16.04.2) but 1.13.1-0ubuntu6 is to be installed
E: Unable to correct problems, you have held broken packages.
```

**Comment by flx42** [comment]
@shrx what's your distro?

**Comment by shrx** [comment]
@flx42 Xubuntu 17.10 with 4.15.7 kernel.

**Comment by flx42** [comment]
Sorry, either use 17.10 with the official `docker-ce` package, or wait for ubuntu 18.04

**Comment by shrx** [comment]
Ah, for some reason I couldn't install the required `docker-ce=17.12.1~ce-0~ubuntu` version yesterday, but today it worked.

**Comment by liwzhi** [comment]
sudo yum install -y docker.io nvidia-container-runtime=1.1.1+docker1.13.1-1 nvidia-docker2=2.0.2+docker1.13.1-1
Loaded plugins: fastestmirror, langpacks
Loading mirror speeds from cached hostfile
No package docker.io available.
No package nvidia-container-runtime=1.1.1+docker1.13.1-1 available.
No package nvidia-docker2=2.0.2+docker1.13.1-1 available.
Error: Nothing to do

**Comment by eaglet3d** [comment]
I am very confused by the nvidia-docker2 documentation.  It specifically says in the pre-requisites that "[_all stable releases of docker-ce are supported_](https://github.com/NVIDIA/nvidia-docker/wiki/Frequently-Asked-Questions#which-docker-packages-are-supported)".  However, when I follow the directions on Xubuntu 18.04 with Docker Version 18.03.1-ce ([which is the latest STABLE release of docker-ce](https://docs.docker.com/release-notes/docker-ce/)), I get:
> sudo apt-get install nvidia-docker2
> Reading package lists... Done
> Building dependency tree       
> Reading state information... Done
> Some packages could not be installed. This may mean that you have
> requested an impossible situation or if you are using the unstable
> distribution that some required packages have not yet been created
> or been moved out of Incoming.
> The following information may help to resolve the situation:
> 
> The following packages have unmet dependencies:
> nvidia-docker2 : Depends: docker.io (= 17.12.1-0ubuntu1) but it is not going to be installed
> E: Unable to correct problems, you have held broken packages.

Are the instructions wrong?  ... or should this ticket stay open until the nvidia-docker2 package can be installed with the latest stable version of docker-ce on Ubuntu 18.04?

**Comment by flx42** [comment]
@eaglet3d The `docker-ce` repository for bionic is currently empty. Only `docker.io` is available (from Canonical)

**Comment by eaglet3d** [comment]
The instructions do not make it clear that only docker.io is supported when using Ubuntu.  Docker Version 18.03.1-ce works on Ubuntu 18.04 when one installs manually.  Therefore, nvidia-docker2's statement that it works with the pre-requisite of "All the stable releases of docker-ce." could be clarified with a table of supported configurations showing: nvidia-docker2 version, OS, OS Version, Docker, Docker Version.

Can someone please clarify what IS supported by fixing the instructions? Otherwise, please re-open this ticket.

**Comment by flx42** [comment]
> The instructions do not make it clear that only docker.io is supported when using Ubuntu.

Ubuntu 18.04

> Docker Version 18.03.1-ce works on Ubuntu 18.04 when one installs manually.

You are using the xenial repository, right?

> "All the stable releases of docker-ce."

It works with all stable releases of `docker-ce` for Ubuntu 18.04, that is the empty set :)

**Comment by eaglet3d** [comment]
The installation documentation of nvidia-docker2 is unclear.  It could be
made better so people don't waste their time with unsupported
configurations, or you can continue to be cute and ignore people's pain
with trying to get nvidia-docker2 working.

Kind regards,
Ralph

Ralph A. Navarro Jr.
*Navarro Computing LLC <http://navarrocomputing.com/home>*
Mobile: *+1(508)**287-0190*
AIM,YIM,Skype: eaglet3d

**Comment by flx42** [comment]
Enlighten me, how did you install `docker-ce`?

**Comment by flx42** [comment]
Added the following in the FAQ:
> The exact list depends on what is available for your Linux distribution.
You can look at the packages list on the gh-pages branch of the repository

**Comment by eaglet3d** [comment]
Thank you!  The change to thehelps to clarify further.

To answer your question, I was able to

**Comment by rockper** [comment]
nvidia-docker2 is working on Ubuntu 18.04, with docker.io from the Ubuntu repo, not docker-ce from docker.com.

**Comment by rockper** [comment]
nvidia-docker2 is working on Ubuntu 18.04, with docker.io from the Ubuntu repo, not docker-ce from docker.com.

**Comment by rockper** [comment]
nvidia-docker2 is working on Ubuntu 18.04, with docker.io from the Ubuntu repo, not docker-ce from docker.com. Use "docker.io", not "docker" which is a snap package  
`$ sudo docker version`

If you have docker-ce, purge it.   
`$ sudo apt-get purge -y docker-ce`  

Edit /etc/apt/sources.list & comment out the line  
\# deb [arch=amd64] https://download.docker.com/linux/ubuntu artful stable

`$ sudo apt-get install docker.io`  
Then continue with the instructions on https://github.com/NVIDIA/nvidia-docker

**Comment by sgregnt** [comment]
Running: 

     sudo apt-get install -y nvidia-docker2

I get the following error:

     The following packages have unmet dependencies:
     nvidia-docker2 : Depends: docker-ce (= 18.03.1~ce~3-0~ubuntu) but 
                        18.05.0~ce~3-0~ubuntu is to be installed or
                       docker-ee (= 18.03.1~ee~3-0~ubuntu) but it is not 

My bionic ubuntu 

        Linux lab1-pc 4.15.0-23-generic #25-Ubuntu SMP Wed May 23 18:02:16 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

I'm new to docker and have a limited experience working with Linux. 
Trying to ramp up git project (a machine learning project hosted on Github). The project uses GPU (nvidia) and relies on docker and nvidia-docker2 and is coded in python (3.5).  Probably due to lacking expirience with linux I can't get the bottom line of this issue, is nvidia-docke2 already supported on ubuntu 18.04 or not yet? 

The work around suggested by rockper did not work for me, I purged docker-ce and installed docker.io: 

     sudo docker version


     Client:
     Version:    17.12.1-ce
     API version:    1.35
     Go version:    go1.10.1
      Git commit:    7390fc6
      Built:    Wed Apr 18 01:23:11 2018
      OS/Arch:    linux/amd64

     Server:
      Engine:
      Version:    17.12.1-ce
      API version:    1.35 (minimum version 1.12)
      Go version:    go1.10.1
      Git commit:    7390fc6
      Built:    Wed Feb 28 17:46:05 2018
      OS/Arch:    linux/amd64
      Experimental:    false

Which is the latest [`docker.io` package on ubuntu site ](https://packages.ubuntu.com/bionic/docker.io) 
Then running 

     sudo apt-get install -y nvidia-docker2

I still get an error

     The following packages have unmet dependencies:
     nvidia-docker2 : Depends: docker-ce (= 18.03.1~ce~3-0~ubuntu) but it is not 
     going to be installed or
                           docker-ee (= 18.03.1~ee~3-0~ubuntu) but it is not installable
     E: Unable to correct problems, you have held broken packages.

The complete code of READM.md that I try to ramp up is [here](https://gist.github.com/sgregnt/cae3a5081561bca7a5cbd0058c8076f4). Any comments or suggestions are welcom.

**Comment by rockper** [comment]
`Version:    17.12.1-ce`  
You have docker-ce from docker.com, not docker.io. 18.03.1-ce has recently become available for ubuntu 18.04 [docker-ce bionic stable](https://download.docker.com/linux/ubuntu/dists/bionic/pool/stable/amd64/).  

Edit /etc/apt/sources.list Add the line below (or if it exists, change xenial/zesty/artful to bionic) 
`deb [arch=amd64] https://download.docker.com/linux/ubuntu bionic stable`  

Save the file above. Then
`sudo apt-get update`  
`sudo apt-get install docker-ce`  

Docker should now be version 18.03.1-ce
`sudo docker version`

Then continue with the instructions on https://github.com/NVIDIA/nvidia-docker

**Comment by sgregnt** [comment]
Thanks for a prompt reply! I think I did install `docker.io` but somehow it shows up as `17.12.1-ce`. Anyway, I did run what you've suggested. But it didn't work I still get 

       The following packages have unmet dependencies:
       nvidia-docker2 : Depends: docker-ce (= 18.03.1~ce~3-0~ubuntu) but 18.05.0~ce~3-0~ubuntu is to be 
       installed or
                           docker-ee (= 18.03.1~ee~3-0~ubuntu) but it is not installable

[here](https://gist.github.com/tzahishimkin/ad193d3ae4a797ec2d5bb2b8d87b22c6) contents of shell session.

**Comment by flx42** [comment]
@sgregnt remove any reference to the `edge` channel of docker in `/etc/apt/sources.list`

**Comment by rockper** [comment]
Your current version of docker-ce is 18.05.0~ce~3-0~ubuntu. You might be installing docker-ce from ["bionic edge"](https://download.docker.com/linux/ubuntu/dists/bionic/pool/edge/amd64/) instead of ["bionic stable"](https://download.docker.com/linux/ubuntu/dists/bionic/pool/stable/amd64/).

In `/etc/apt/sources.list` make sure the last word is `stable`, not `edge`
`deb [arch=amd64] https://download.docker.com/linux/ubuntu bionic stable`

Then the usual:
`sudo apt-get update`
`sudo apt-get install -y docker-ce`

Verify your docker version with:
`sudo docker version`

nvidia-docker2 requires a specific version of docker, 18.03.1~ce~3-0~ubuntu.
PS: [docker.io](https://packages.ubuntu.com/bionic/docker.io) from the Ubuntu repo is no longer required, because docker-ce is now available for bionic

**Comment by hamzafar** [comment]
Hi @flx42 ,

I am using ubuntu 16.04, and connect install nvidia-docker2. Could you please recommend me either the   nvidia-docker2 can be installed on ubuntu 16.04 or we need to update to ubuntu 17.10.

But, our perferred ubuntu version is 16.04.

Regards
Hamza Zafar

**Comment by flx42** [comment]
@hamzafar You should be able to install `nvidia-docker2` for Ubuntu 16.04. If that's not the case please open a new issue.

**Comment by mysablehats** [comment]
What I did was remove the version of docker nvidia was having a problem with:
```
sudo apt remove docker-ce
sudo apt autoremove
```
Then I installed the version it was asking for:
` sudo apt-get install docker-ce=18.06.1~ce~3-0~ubuntu`

This fixed the issue for me, but well, I suppose this should 'just work'

**Comment by jmaspons** [comment]
> ```
>  sudo apt-get install -y nvidia-docker2
> ```
> 
> I get the following error:
> 
> ```
>  The following packages have unmet dependencies:
>  nvidia-docker2 : Depends: docker-ce (= 18.03.1~ce~3-0~ubuntu) but 
>                     18.05.0~ce~3-0~ubuntu is to be installed or
>                    docker-ee (= 18.03.1~ee~3-0~ubuntu) but it is not 
> ```

I have the same problem in ubuntu bionic. I see how to install docker-ce but I prefer to use docker.io. Could you add docker.io as an alternative dependency to docker-ce?

FAQS says that docker.io is also supported https://github.com/NVIDIA/nvidia-docker/wiki/Frequently-Asked-Questions#which-docker-packages-are-supported

**Comment by christk1** [comment]
> What I did was remove the version of docker nvidia was having a problem with:
> 
> ```
> sudo apt remove docker-ce
> sudo apt autoremove
> ```
> 
> Then I installed the version it was asking for:
> ` sudo apt-get install docker-ce=18.06.1~ce~3-0~ubuntu`
> 
> And then proceeded to install nvidia-docker2
> This fixed the issue for me, but well, I suppose this should 'just work'

sometimes the answer is in front of your eyes but you cant see it :P

**Comment by AndreV84** [comment]
now it seems to be:

> sudo apt install nvidia-docker2
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 nvidia-docker2 : Depends: docker-ce (= 5:18.09.0~3-0~ubuntu-bionic) but 5:18.09.1~3-0~ubuntu-bionic is to be installed or
                           docker-ee (= 5:18.09.0~3-0~ubuntu-bionic) but it is not installable


sudo apt-get install docker-ce=5:18.09.0~3-0~ubuntu-bionic

> apt list --upgradable -a
Listing... Done
docker-ce/bionic 5:18.09.1~3-0~ubuntu-bionic amd64 [upgradable from: 18.06.1~ce~3-0~ubuntu]
docker-ce/bionic 5:18.09.0~3-0~ubuntu-bionic amd64
docker-ce/bionic,now 18.06.1~ce~3-0~ubuntu amd64 [installed,upgradable to: 5:18.09.1~3-0~ubuntu-bionic]
docker-ce/bionic 18.06.0~ce~3-0~ubuntu amd64
docker-ce/bionic 18.03.1~ce~3-0~ubuntu amd64

> > Setting up docker-ce (5:18.09.0~3-0~ubuntu-bionic) ...
update-alternatives: using /usr/bin/dockerd-ce to provide /usr/bin/dockerd (dockerd) in auto mode
Job for docker.service failed because the control process exited with error code.
See "systemctl status docker.service" and "journalctl -xe" for details.

**Comment by AndreV84** [comment]
what worked was:
`  sudo apt-get install docker-ce=5:18.09.0~3-0~ubuntu-bionic`
and overriding ' sudo systemctl edit docker.service'
with

>  [Service]
> ExecStart=
> ExecStart=/usr/bin/dockerd -H unix:// -H tcp://127.0.0.1:2375

**Comment by AndreV84** [comment]
@herbert-attila-verizon : Yes, I do. The information from previous [attempts] in the file contained conflicted parameters. smth. like nvidia-runtime=...

**Comment by richardrl** [comment]
> what worked was:
> ` sudo apt-get install docker-ce=5:18.09.0~3-0~ubuntu-bionic`
> and overriding ' sudo systemctl edit docker.service'
> with
> 
> > [Service]
> > ExecStart=
> > ExecStart=/usr/bin/dockerd -H unix:// -H tcp://127.0.0.1:2375

Do you have to upgrade 18.04 to 18.09 exactly to do this? Currently enable to get things running with 18.04
